src/payload.rs

```diff
@@ -158,6 +163,59 @@ pub unsafe extern "C" fn z_bytes_decode_into_slice(
     }
 }
 
+#[cfg(all(feature = "shared-memory", feature = "unstable"))]
+/// Decodes data into an owned SHM buffer by copying it's shared reference
+///
+/// @param this_: Data to decode.
+/// @param dst: An unitialized memory location where to construct a decoded string.
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn z_bytes_decode_into_owned_shm(
+    this: &z_loaned_bytes_t,
+    dst: *mut MaybeUninit<z_owned_shm_t>,
+) -> z_error_t {
+    use zenoh::shm::zshm;
+
+    let payload = this.transmute_ref();
+    match payload.deserialize::<&zshm>() {
+        Ok(s) => {
+            Inplace::init(dst.transmute_uninit_ptr(), Some(s.clone().to_owned()));
+            errors::Z_OK
+        }
+        Err(e) => {
+            log::error!("Failed to decode the payload: {}", e);
+            Inplace::empty(dst.transmute_uninit_ptr());
+            errors::Z_EIO
+        }
+    }
+}
+
+#[cfg(all(feature = "shared-memory", feature = "unstable"))]
+/// Decodes data into a loaned SHM buffer
+///
+/// @param this_: Data to decode.
+/// @param dst: An unitialized memory location where to construct a decoded string.
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn z_bytes_decode_into_loaned_shm(
+    this: &z_loaned_bytes_t,
+    dst: *mut MaybeUninit<&'static z_loaned_shm_t>,
```

Comment:
This is an API to reflect Rust's `payload.deserialize::<&zshm>()`. User can get access to loaned SHM buffer through loaned ZBytes without making a copy of SHM buffer.

Reply:
We can return *z_loaned_shm_t as a return value, but in this case we'll anyway need to return nullptr in case of error - and here we are again with uninitialized loaned type.....

Reply:
Maybe we need to trigger some discussion on that

Reply:
strange &zshm means that it is immutable, while so far all decode functions give user an owned object which he is free to mutate as much as he wants. Maybe rename it to something like z_bytes_as_shm() and return (maybe null) pointer directly ?

Reply:
This C API reflects `deserialize<>` method for Rust API, so I think it should look and feel in the same way as anything that  deserializes into owned object.

Another question is that we might want to `deserialize_mut` into a mutably-loaned object, so I'd rather add `z_bytes_decode_into_mut_loaned_shm` call here